### ISSUE TITLE
Timeout waiting for mount pvc in Kubernetes (2)

### ISSUE THREAD

**Issue by nicned** [seaweedfs/seaweedfs-csi-driver#67]
Hi,
I don't open this as a general bug. We run quite a specific setup and I am thinking that the problem is specific to the way we try to run SeaweedFS and Kubernetes. Still, by this, I would like to ask for some help in the troubleshooting. We are quite keen to make it work.

On the surface our issue is very similar to #35. For what I see the similarities with #35 are in the actual error, and in the fact that they've also run Alpine Linux; but for the rest our setup is quite different. Hence, I think we should track this separately from #35 for now. 

----------------------------
We run a 3-node SeaweedFS cluster on ARM64 embedded devices with Alpine Linux. Each node runs a master-, volume-server, a filer and it has local FUSE mounts. We use Redis for SeaweedFS’s database. The bare-metal SeaweedFS cluster works well on the 3 nodes. We don't see any issues there. This is how we run weeds
```
sudo ps -Afw --forest | grep weed
root     15740 15647  0 Jul03 ?        00:00:00      |       |   \_ /csi-node-driver-registrar --v=5 --csi-address=/csi/csi.sock --kubelet-registration-path=/var/lib/kubelet/plugins/seaweedfs-csi-driver/csi.sock
root     16251 16195  0 Jul03 ?        00:00:00      |       |   \_ /seaweedfs-csi-driver --endpoint=unix:///csi/csi.sock --filer=127.0.0.1:5555 --nodeid=e10ccwe080c000001127
root     26071 20654  0 Jul03 ?        00:00:09          \_ ./weed master -mdir=/vmrun/sweeds -defaultReplication=001 -peers=10.101.35.11:9333,10.101.115.39:9333
root     26361 20654  0 Jul03 ?        00:00:01          \_ ./weed volume -dir=/vmrun/sweeds/volume -dataCenter=bath -rack=ate-rack
root     26663 20654  0 Jul03 ?        00:00:01          \_ ./weed filer -master=10.101.0.1:9333 -port=5555
root     28257 20654  0 00:00 ?        00:00:01          \_ ./weed mount -filer=localhost:5555 -dir=/tmp/weed-mount-1 -filer.path=/
```

The same three nodes are also a Kubernetes cluster. We run K3s (the lightweight Kubernetes distribution created by Rancher Labs). The other peculiarity is that, for number of reasons, we use Docker for container runtime. The Kubernetes cluster is also fully functional; we run variety of workloads w/o problems. 
```
NAME                   STATUS   ROLES                  AGE   VERSION        INTERNAL-IP     EXTERNAL-IP   OS-IMAGE            KERNEL-VERSION     CONTAINER-RUNTIME
e10ccwe080c000001127   Ready    control-plane,master   38d   v1.23.6+k3s1   172.31.0.143    <none>        Alpine Linux v3.9   4.9.291-vsys-1.0   docker://18.9.1
e10ctwe080c000002458   Ready    <none>                 16d   v1.23.6+k3s1   10.101.115.39   <none>        Alpine Linux v3.9   4.9.291-vsys-1.0   docker://18.9.1
e09ctwe080b000102186   Ready    <none>                 16d   v1.23.6+k3s1   10.101.35.11    <none>        Alpine Linux v3.9   4.9.291-vsys-1.0   docker://18.9.1

```

We are failing when we try to use SeaweedFS via the SeaweedFS-CSI-driver in Kubernetes. The SeaweedFS CSI Driver seems to install properly. These are the pods.
kubectl get pods -o wide
```
NAME                     READY   STATUS    RESTARTS      AGE     IP           NODE                   NOMINATED NODE   READINESS GATES
seaweedfs-node-xmb6q     2/2     Running   0             2d23h   10.42.0.87   e10ccwe080c000001127   <none>           <none>
seaweedfs-controller-0   4/4     Running   0             2d23h   10.42.2.14   e10ctwe080c000002458   <none>           <none>
seaweedfs-node-wk999     2/2     Running   0             2d23h   10.42.2.15   e10ctwe080c000002458   <none>           <none>
ubuntu-pod               1/1     Running   0             2d22h   10.42.0.88   e10ccwe080c000001127   <none>           <none>
seaweedfs-node-hvl7g     2/2     Running   2 (22h ago)   2d23h   10.42.1.15   e09ctwe080b000102186   <none>           <none>
```

We go for a PersistenVolume. Both the PV and PVC seem to be created and available properly. 

```
kubectl get pv 
NAME               CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                                       STORAGECLASS   REASON   AGE
seaweedfs-static   2Gi        RWX            Retain           Bound    default/seaweedfs-persistent-volume-claim                           3d1h
```

```
kubectl get pvc 
NAME                                STATUS   VOLUME             CAPACITY   ACCESS MODES   STORAGECLASS   AGE
seaweedfs-persistent-volume-claim   Bound    seaweedfs-static   2Gi        RWX                           3d1h
```

However any attempt to use the persistent volume claim in a pod, from your busybox sample or others, fail with the same error
`E0704 00:08:15.376919    7303 nestedpendingoperations.go:335] Operation for "{volumeName:kubernetes.io/csi/seaweedfs-csi-driver^dfs-test podName: nodeName:}" failed. No retries permitted until 2022-07-04 00:08:16.376835824 +0100 BST m=+97464.670614673 (durationBeforeRetry 1s). Error: MountVolume.SetUp failed for volume "seaweedfs-static" (UniqueName: "kubernetes.io/csi/seaweedfs-csi-driver^dfs-test") pod "my-csi-app" (UID: "5afa6d91-10eb-48bb-9cca-6d94c05ac616") : rpc error: code = Internal desc = Timeout waiting for mount`

The error message itself isn't very descriptive. We could guess that some process somewhere times-out, but where, why ??
I could not find anything else in the k3s.log or in the general log, to give me a cue.

Do you have a suggestion where I could go, and where to look from here ?

**Comment by nicned** [comment]
I've just spotted that on the node where the pod with SeaweedFS is scheduled to run, we get tons of new weed processes. 

```
ps -Afw --forest | grep weed
root     17223 17183  0 Jul02 ?        00:00:10              |   \_ /csi-node-driver-registrar --v=5 --csi-address=/csi/csi.sock --kubelet-registration-path=/var/lib/kubelet/plugins/seaweedfs-csi-driver/csi.sock
root     17418 17355  0 Jul02 ?        00:00:15              |   \_ /seaweedfs-csi-driver --endpoint=unix:///csi/csi.sock --filer=127.0.0.1:5555 --nodeid=e09ctwe080b000102186
root     11457 17418  0 00:07 ?        00:00:00              |       \_ [weed] <defunct>
root     11779 17418  0 00:08 ?        00:00:00              |       \_ [weed] <defunct>
root     12127 17418  0 00:08 ?        00:00:00              |       \_ [weed] <defunct>
root     12501 17418  0 00:08 ?        00:00:00              |       \_ [weed] <defunct>
root     12953 17418  0 00:08 ?        00:00:00              |       \_ [weed] <defunct>
root     13505 17418  0 00:09 ?        00:00:00              |       \_ [weed] <defunct>
root     14336 17418  0 00:09 ?        00:00:00              |       \_ [weed] <defunct>
root     15645 17418  0 00:10 ?        00:00:00              |       \_ [weed] <defunct>
root     18056 17418  0 00:11 ?        00:00:00              |       \_ [weed] <defunct>
root     22031 17418  0 00:13 ?        00:00:00              |       \_ [weed] <defunct>
root     26132 17418  0 00:15 ?        00:00:00              |       \_ [weed] <defunct>
root     30033 17418  0 00:18 ?        00:00:00              |       \_ [weed] <defunct>
root      1569 17418  0 00:20 ?        00:00:00              |       \_ [weed] <defunct>
root      5600 17418  0 00:22 ?        00:00:00              |       \_ [weed] <defunct>
root      9681 17418  0 00:24 ?        00:00:00              |       \_ [weed] <defunct>
root     13610 17418  0 00:26 ?        00:00:00              |       \_ [weed] <defunct>
root     17760 17418  0 00:29 ?        00:00:00              |       \_ [weed] <defunct>
root     21857 17418  0 00:31 ?        00:00:00              |       \_ [weed] <defunct>
root     25811 17418  0 00:33 ?        00:00:00              |       \_ [weed] <defunct>
root     29730 17418  0 00:35 ?        00:00:00              |       \_ [weed] <defunct>
root      1197 17418  0 00:37 ?        00:00:00              |       \_ [weed] <defunct>
root      5241 17418  0 00:40 ?        00:00:00              |       \_ [weed] <defunct>
root      9319 17418  0 00:42 ?        00:00:00              |       \_ [weed] <defunct>
root     13287 17418  0 00:44 ?        00:00:00              |       \_ [weed] <defunct>
root     17436 17418  0 00:46 ?        00:00:00              |       \_ weed -logtostderr=true mount -dirAutoCreate=true -umask=000 -dir=/var/lib/kubelet/pods/5afa6d91-10eb-48bb-9cca-6d94c05ac616/volumes/kubernetes.io~csi/seaweedfs-static/mount -collection=default -collectionQuotaMB=0 -filer=127.0.0.1:5555 -filer.path=/vmrun/sweeds/volume -cacheCapacityMB=0 -localSocket=/tmp/seaweedfs-mount-1470629645.sock -replication=001 -concurrentWriters=32 -cacheDir=/tmp
root     13765 24559  0 Jul03 ?        00:02:40          \_ ./weed master -mdir=/vmrun/sweeds -defaultReplication=001 -peers=10.101.0.1:9333,10.101.115.39:9333
root     14092 24559  0 Jul03 ?        00:00:14          \_ ./weed volume -dir=/vmrun/sweeds/volume -dataCenter=bath -rack=ate-rack
root     14907 24559  0 Jul03 ?        00:00:28          \_ ./weed filer -port=5555
root     22718 24559  0 Jul03 ?        00:00:13          \_ ./weed mount -filer=localhost:5555 -dir=/tmp/weed-mount-1 -filer.path=/
```
...not sure if this helps.

**Comment by chrislusf** [comment]
Maybe try a different cis-driver version via git bisect.

**Comment by nicned** [comment]
Hi Chris,
I noticed in the process list above (with the many [weed] <defunct>) the process 17436.
```
weed -logtostderr=true mount -dirAutoCreate=true -umask=000 -dir=/var/lib/kubelet/pods/5afa6d91-10eb-48bb-9cca-6d94c05ac616/volumes/kubernetes.io~csi/seaweedfs-static/mount -collection=default -collectionQuotaMB=0 -filer=127.0.0.1:5555 -filer.path=/vmrun/sweeds/volume -cacheCapacityMB=0 -localSocket=/tmp/seaweedfs-mount-1470629645.sock -replication=001 -concurrentWriters=32 -cacheDir=/tmp
root     13765 24559  0 Jul03 ?        00:02:40          \_ ./weed master -mdir=/vmrun/sweeds -defaultReplication=001 -peers=10.101.0.1:9333,10.101.115.39:9333
```

I try to run that command manually, and exactly as in #35 I get
`can not stat /var/lib/kubelet/pods/5afa6d91-10eb-48bb-9cca-6d94c05ac616/volumes/kubernetes.io~csi/seaweedfs-static/mount`

and exactly as in #35 we are missing the part after /var/lib/kubelet/pods/5afa6d91-10eb-48bb-9cca-6d94c05ac616/volumes/kubernetes.io~csi/

Our issue seem to have much more in common with #35 that I thought originally. Did you manage to find before why these folders are not created?   ...I wonder if there is something to do with Alpine Linux.

I created manually the missing folders `seaweedfs-static `and `mount`, but I get the same error when I try to run the above weed command line manually.

BTW, should mount be a folder or a file ?

**Comment by nicned** [comment]
Could you please tell me also the purpose of the `-filer.path=/vmrun/sweeds/volume` option in the above command?

I realize this is taken from the persistent-volume yaml
```
$cat seaweed-csi-persistent-volume.yaml 
apiVersion: v1
kind: PersistentVolume
metadata:
  name: seaweedfs-static
spec:
  storageClassName: ""
  accessModes:
  - ReadWriteMany
  capacity:
    storage: 2Gi
  csi:
    driver: seaweedfs-csi-driver
    volumeHandle: dfs-test
    volumeAttributes:
      collection: default
      replication: "001"
      path: /vmrun/sweeds/volume
   #   diskType: "ssd"
   # readOnly: true
  persistentVolumeReclaimPolicy: Retain
  volumeMode: Filesystem
```
But I noticed that this folder does exist either on my node.

I made a new experiment. I commented out that particular line, i.e. it takes the default `/buckets`. The option in the weed process changed accordingly, i.e. it's now
```
./weed -logtostderr=true mount -dirAutoCreate=true -umask=000 -dir=/var/lib/kubelet/pods/fe1897da-7b29-48eb-817c-d2d
f64bfc104/volumes/kubernetes.io~csi/seaweedfs-static/mount -collection=default -collectionQuotaMB=0 -filer=127.0.0.1:5555 -filer.path=/bucke
ts/dfs-test -cacheCapacityMB=0 -localSocket=/tmp/seaweedfs-mount-1470629645.sock -replication=001 -concurrentWriters=32 -cacheDir=/tmp
```
 ...but the node where the pod is scheduled to run doesn't have a folder `/buckets`. It should, right?

I wonder if we have some general 'new folder creation issue' with the SeaweedFS CSI driver in Alpine Linux.  Please let me know your thoughts.

**Comment by nicned** [comment]
We made an interesting discovery today. 
If we execute manually the `weed `command that the code is trying to run, we get exactly the same error as the gentlemen in #35 
```
~$ ./weed -logtostderr=true mount -dirAutoCreate=true -umask=000 -dir=/var/lib/kubelet/pods/c233def2-5a4e-43ad-a862-f8d
b1e56687e/volumes/kubernetes.io~csi/seaweedfs-static/mount -collection=default -collectionQuotaMB=0 -filer=127.0.0.1:5555 -filer.path=/bucke
ts/dfs-test -cacheCapacityMB=0 -localSocket=/tmp/seaweedfs-mount-1470629645.sock -replication=001 -concurrentWriters=32 -cacheDir=/tmp
can not stat /var/lib/kubelet/pods/c233def2-5a4e-43ad-a862-f8db1e56687e/volumes/kubernetes.io~csi/seaweedfs-static/mount
```
And, the error is because the last two folders in the path `seaweedfs-static` and `mount` do not exist. Even if we create them manually, the error `can not stat ...` persists.

HOWEVER, if we just `sudo `the same `weed` command, it creates properly the folders in the path and it executes successfully 

```
$sudo ./weed -logtostderr=true mount -dirAutoCreate=true -umask=000 -dir=/var/lib/kubelet/pods/c233def2-5a4e-43ad-a86
2-f8db1e56687e/volumes/kubernetes.io~csi/seaweedfs-static/mount -collection=default -collectionQuotaMB=0 -filer=127.0.0.1:5555 -filer.path=/
buckets/dfs-test -cacheCapacityMB=0 -localSocket=/tmp/seaweedfs-mount-1470629645.sock -replication=001 -concurrentWriters=32 -cacheDir=/tmp 
&
E10CCWE080C000001127:~$ mount point owner uid=0 gid=0 mode=drwxrwxrwx
current uid=0 gid=0
I0704 22:05:28  3440 leveldb_store.go:47] filer store dir: /tmp/c12d5add/meta
I0704 22:05:28  3440 file_util.go:23] Folder /tmp/c12d5add/meta Permission: -rwxr-xr-x
This is SeaweedFS version 30GB 3.02 a4ca3ed1f0c2bcde1d36bb48f81dedc3e1924679 linux arm64
```
...and the respective pod springs to life!

There is apparently some privileges/rights issue. It seem that the code (or kubelet) is somehow executing with insufficient privileges.  I don't quite understand why is that. Any idea where the issue might be?

**Comment by nicned** [comment]
I am closing the issue because we found and fixed the problem. 
It was caused by a configuration error on our side. We were using 127.0.0.1 (localhost) as Filer's address, i.e.
`SEAWEEDFS_FILER` had `value: "127.0.0.1:5555"` in our seaweedfs-csi.yaml.  And the weed commands were failing because the CSI driver runs in a pod/container, where the meaning of 127.0.0.1 is different (than on the host).

We were using the localhost's IP (127.0.0.1) in a first place because we thought this is how we could achieve redundancy between multiple Filers. We have a three node cluster and each node runs a Filer. Obviously setting one Filer's IP addess in `SEAWEEDFS_FILER` does not provide a redundancy when that Filer goes offline.

**The correct solution** is to provide all Filers' addresses as a comma-separated-list for `SEAWEEDFS_FILER`, e.g.
```
- name: SEAWEEDFS_FILER
              value: "10.101.115.39:5555,10.101.0.1:5555,10.101.35.11:5555"
```

P.S.  It may be a good idea to add a note about this to the main README.

Note as well that the list does not tolerate empty spaces, i.e.  `value: "10.101.115.39:5555, 10.101.0.1:5555" `will fail; meaning the second IP would not be taken into consideration.
